Fixes: Library - Connection state's position when there are no libraries around #11442

yashhash2 opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
So I created media queries for specific screen sizes to display the connection state position correctly. applied Padding and margins wherever necessary and created 3 DIVS a, b and disco because when state was offline position was changed again,by using these DIVS I was able to fix issue for both offline and online state.




## References

https://github.com/learningequality/kolibri/issues/11442#issue-1954449606

yashhash2 committed:
Your commit message

MisRob reviewed (dismissed):
Hi @yashhash2, thank you. High-level feedback from me:

- Would you upload before/after screenshots to the pull request description?
- Generally, we [don't use CSS media queries](https://kolibri-dev.readthedocs.io/en/develop/frontend_architecture/conventions.html#styling-anti-patterns). Instead we utilize [useKResponsiveWindow](https://design-system.learningequality.org/usekresponsivewindow/) that is integrated with [our breakpoint system.](https://design-system.learningequality.org/layout/#responsiveness) You could read the docs and study another places in the codebase. Then I'd ask you to remove the CSS media queries in favor of `useKResponsiveWindow`.
- I see many duplicate styles. Would you place styles common for all breakpoints outside particular breakpoint styles so they are applied to all breakpoints? Then from each breakpoint style only override what is needed for that particular breakpoint.  That should simplify code significantly.

MisRob commented:
I will invite my colleague @LianaHarris360 for a full review as soon as this is resolved. Liana, would you please have a look then? I only did a brief skim myself. This is a new attempt at https://github.com/learningequality/kolibri/issues/11442 which you reviewed once here https://github.com/learningequality/kolibri/pull/12405.

yashhash2 commented:
Thank you @MisRob for feedback i will try to incorporate as many points as possible.

yashhash2 committed:
Your commit message

yashhash2 commented:
![Screenshot from 2024-12-17 17-21-15](https://github.com/user-attachments/assets/a6dd4130-feb1-4fd4-8889-c56d83150b0a)

This is SS of the change, I have also added the changes in ordered manner Please look into it. I tried using Kresponsive Element but still i have to add css at breakpoints because of logo there its overlapping with text in some screen sizes.

yashhash2 commented:
@LianaHarris360 Please Review the changes so I can do any other change if required

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -13,12 +14,7 @@
           {{ injectedtr('otherLibraries') }}
         </h1>
-      </KGridItem>
-      <KGridItem
```
Removing the usage of `KGridItem` here or the `margin-left: -190px;` on line 322 might have had unwanted effects on the icon and string shown when searching for other libraries:

![searchingLibraries](https://github.com/user-attachments/assets/7e1abae3-08a3-49e8-becc-5889e1004754)


I also noticed that when the window is large, the `No other libraries around you right now` and `Showing all available libraries around you` descriptions are no longer positioned underneath the `Other Libraries` section header:

![noOtherLibraries](https://github.com/user-attachments/assets/b3b3b593-c11e-4bd3-9ca1-dac7bd90318b)
![availableLibraries](https://github.com/user-attachments/assets/d08a8be1-6df5-4a15-b98e-bf8466099fba)

I think any changes you make to the connection state when there are no libraries should also be compared with the other two connection state descriptions to make sure there are no regressions.

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -257,4 +286,67 @@
     margin-left: 0.75em;
   }
 
+  @media screen and (max-width: 600px) {
```
`useKResponsiveWindow` is used for styling based on reactive window size information rather than CSS Media queries. For more detailed breakpoint styling, using `KGridItem` along with `useKResponsiveWindow` could be useful. There’s some examples of this being done in the Kolibri codebase [here](https://github.com/yashhash2/kolibri/blob/48eaf0366de77bb480dd5fce726433932845707d/kolibri/plugins/facility/assets/src/views/DataPage/index.vue#L109), [here](https://github.com/yashhash2/kolibri/blob/my-awesome-changes/kolibri/plugins/device/assets/src/views/HeaderWithOptions.vue#L15), and [here](https://github.com/yashhash2/kolibri/blob/48eaf0366de77bb480dd5fce726433932845707d/kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/QuestionListPreview.vue#L5).

I saw that you mentioned logo overlapping with text on some screen sizes; perhaps this issue could be fixed by using conditional styling?

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -55,16 +51,25 @@
+          <div
             v-show="!searchingOtherLibraries && !devicesWithChannelsExist"
-            data-test="no-other"
+            class="a"
```
CSS class names are most effective when they are descriptive and meaningful, and indicate the purpose of the element. It would be best to avoid using vague or generic names. 

Maybe instead of class names `a` and `b`, something like `no-other-div` and `no-other-label-div`? This will also make the CSS easier to read and helps other devs quickly identify which element it is referring to.

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -257,4 +286,67 @@
+      }
+
+      .disco {
+        margin-right: -640px;
```
I noticed more than 10 instances of negative margins being used. While they can be useful at times, they can be difficult to debug and make the CSS harder to read, so we try to avoid them or add a comment explaining its usage. Is there an alternative you can use in place of some of these?

LianaHarris360 reviewed (changes requested):
Thanks so much for your work on this @yashhash2, this is a tricky issue to fix. There were a few things I pointed out in my review. Please be aware that using CSS media queries is not a recommended practice in the Kolibri codebase. Although it is a straightforward fix, it can lead to challenges with maintenance, and there is a well-established alternative that we use consistently throughout the codebase.

yashhash2 commented:
Thank you for your feedback @LianaHarris360 I will try to do all the changes you mentioned above ASAP

LianaHarris360 commented:
There's no hurry :) Since Learning Equality will be closing for the holidays soon, I might not be able to review more changes until after the new year. Thank you for your efforts!

yashhash2 commented:
Hey @LianaHarris360 I tried using KGridItem instead of CSS media queries and here are some issues in which I need clarifications
Issues Encountered with KGridItem Implementation
Context:
While implementing KGridItem in place of CSS media queries, I observed several challenges regarding layout responsiveness and alignment. Below are the specific issues that require clarification and assistance:

1. Text Alignment Above 600px (Avoiding Specific Colors):
The text renders correctly for screen sizes greater than 600px, resembling the example screenshot provided below. However, the use of blue and pink colors in these tests was solely to visualize exact div dimensions and should be excluded in the final design.
![Screenshot from 2024-12-28 21-41-01](https://github.com/user-attachments/assets/2bb2e980-e25a-422b-91f7-23058d655ad8)

2. Transition Issues Below 600px:
For screen sizes just below 600px (e.g., 550px), the layout undergoes an unintended transition between two shapes or configurations.
![Screenshot from 2024-12-28 21-41-35](https://github.com/user-attachments/assets/18a3848b-2825-4967-9ce9-e51fd8e46f97)
After refreshing, the issue sometimes resolves but occasionally leads to misaligned or out-of-place elements.
At smaller sizes, the problem disappears around 515px, where the layout stabilizes.
Notable Example: The problem occurs at 550px, yet at 515px, it corrects itself as shown:
![Screenshot from 2024-12-28 21-41-49](https://github.com/user-attachments/assets/2c2dbf6c-3e0a-420e-9ec8-231753779b3c)

Key Concern: Despite testing multiple span sizes within the layout configuration (in layout4), this transition anomaly persists. Guidance on resolving this is essential.

3. Wifi Icon Position Below 400px:
At very small screen sizes (generally below 400px), the Wi-Fi icon shifts from its intended right-side position to the left side of the layout.

Example: Wi-Fi icon misplacement below 400px:
![Screenshot from 2024-12-28 21-42-20](https://github.com/user-attachments/assets/73ef266d-a3be-4689-b048-6d55f0182361)
Goal: Resolve this issue while avoiding the use of CSS media queries.

Questions:
How can I prevent the layout transition issue just below 600px, especially ensuring stable behavior across the problematic breakpoints (e.g., 550px to 515px)?
What configurations or adjustments in KGridItem can prevent the Wi-Fi icon from shifting positions below 400px?
Are there recommended best practices for using KGridItem to manage dynamic layouts without media queries effectively?
Thank you for your assistance!

LianaHarris360 commented:
Hi @yashhash2 It would be helpful if I could see the changes you've made in the code; however, I do see that the “Other Libraries” label and the sync-status div are positioned in two KGridItem components next to each other.

However, with the way KGridItem operates, because both KGridItem components have a span of 6 set for the `:layout12` setting, they are displayed side by side. But we want the second KGridItem to be positioned below the first KGridItem on large and small screen sizes.

Perhaps if the span for the KGridItems were increased to numbers greater than 6 in `layout12` and greater than 4 in `layout8`, the KGridItems would stack as intended?

As for the position of the Wi-Fi icon on small screens, it looks like that placement is done manually in the code. I'm not exactly sure why it's set up this way, but I'll investigate to find out the reason. However, since this was done on purpose and is not a result of your changes, you don't need to worry about it!

yashhash2 committed:
added K-grid

yashhash2 commented:
Hi @LianaHarris360,

I've committed the code for you to review. If you'd like me to revert the span of the other-Libraries grid to its original size, I can make that adjustment as well.

Please note that the breaking point is still the primary concern, and the current layout places the K-grid items below one another due to their span sizes exceeding 6 and 4, respectively. Let me know your thoughts or if you have additional feedback.

Best regards,
Yash Kumar Singh

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -4,21 +4,23 @@
+
+        <KGridItem
+        :layout12="{ span:8 }"
+        :layout8="{ span:5 }"
```
On larger screens, the connection status text looks fine, but it shrinks on smaller displays even when there's enough space. I would suggest using sizes 8 and 12 to make the most of the available space for the text.

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -233,14 +239,15 @@
     display: flex;
     justify-content: flex-end;
-    margin: 30px 0 10px;
+    margin: 30px 0 0px;
```
The bottom margin was previously set to 10px rather than 0px, which provides more space below the connection status text. 

However, since there is already a bottom margin of 20px in the `.connection-status` class, the total bottom margin is actually 30px. I recommend putting the bottom margin in just one place, either here or in the `.connection-status` class.

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -96,7 +102,7 @@
     >
       <KGridItem
-        :layout12="{ span: 10 }"
+        :layout12="{ span: 6 }"
```
This `KGridItem` is in a separate `KGrid` from the ones above, was there any particular reason for changing the span from 10 to 6?

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -257,4 +264,21 @@
+  }
+  .otherLib{
+    margin-top: -30px;
+    margin-right: 170px;
```
I believe the `flex-end` on the `.sync-status ` is pushing the text over to the right. If you remove `justify-content: flex-end;` on line 244, you shouldn't need the `margin-right` and `padding-right` for these two classes anymore.

In addition, the negative `margin-top` in these two classes conflicts with the `margin: 30px` in the `.sync-status` class and the top margin in the `.connection-status` class. 

The h1 above the status already has a `margin-bottom` that creates space between the **Other libraries** text and the connection status text. If the top margin in `.sync-status` were removed, would the top margins in `.connection-status`, `.noOtherLib`, and `.otherLib` still be needed? 

![sync-statusclass](https://github.com/user-attachments/assets/5e1443c2-a033-434a-97be-e7bef8719cba)

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -257,4 +264,21 @@
+    margin-bottom:20px;
+    padding-top: 0px;
+    margin-left:  -8px;
+    margin-right: 250px;
```
This class has a lot of styling that we might not need, besides maybe the `margin-bottom` and the `margin-left`.

If the `justify-content: flex-end` is removed from `.sync-status`, this `margin-right` could be removed.

Additionally, the negative `margin-top` is also in conflict with the top margin set in `.sync-status` and I don't think both are necessary, perhaps neither. 

The image below shows the dimensions of this `KGridItem` and it appears that there is no padding set on it, so the `padding-top` could be removed as well.


![connection-status](https://github.com/user-attachments/assets/97a3480b-4c70-451e-ac38-11f9367c246b)

LianaHarris360 reviewed (changes requested):
Thanks for making the suggested changes and updating the grid span sizes, we are making great progress! I noted a few things that could improve the spacing of the connection status test.

yashhash2 commented:
Thank You Lana Harris for the Review I will do the recommended changes ASAP : )

yashhash2 committed:
removed unnecessary styles

yashhash2 commented:
Hi @LianaHarris360,

I wanted to let you know that we have successfully resolved the issues we discussed earlier. The following changes were made:

Removed the no-other-Libraries and Other-libraries sections.
Adjusted the right margins and consolidated all the top and bottom margins within the connection-status section.
These changes have effectively fixed the issue. Thank you for your assistance and guidance in resolving this.

Please review the changes at your convenience. Below is the final result for your reference.
![Screenshot from 2025-01-04 14-18-24](https://github.com/user-attachments/assets/fffdf2a3-f4eb-42ba-8055-1672d23a3f98)

Best regards,
Yash Kumar Singh

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -63,10 +65,12 @@
           </span>
         </div>
       </KGridItem>
+    </KGridItem>
```
It seems that this `KGridItem` is placed within another one, but the otherLibraries text should be within its own `KGridItem`, separate from the grid where the connection status is located, like this:
```
<KGridItem>
  <h1> otherLibraries text </h1>
</KGridItem>
<KGridItem>
  <div class="sync-status">
  </div>
</KGridItem>
```

LianaHarris360 commented on kolibri/plugins/learn/assets/src/views/LibraryPage/OtherLibraries.vue:
```diff
@@ -257,4 +260,11 @@
-</style>
+  .connection-status{
+    margin-bottom: 30px;
+    margin-top: -15px;
```
The negative `margin-top` here combined with the negative `margin-bottom` on line 7 makes the connection status text too close to the text above it and is inconsistent with the spacing of the text below it, removing the negative margins and perhaps making the `margin-bottom` in this class smaller should even it out.

![ConnectionStatusSpacing](https://github.com/user-attachments/assets/637e5a74-9981-4341-9cd8-9a20de8481c3)

yashhash2 committed:
fixed k-grid component

yashhash2 committed:
Update index.vue

yashhash2 commented:
Hi @LianaHarris360 Sorry for dragging this issue to this far. I wasnt sure about the exact spacing which is required so here is the final preview of the current spacing.
![Screenshot from 2025-01-07 17-25-54](https://github.com/user-attachments/assets/1ebdcee5-72f7-472e-ac1a-574b7d8d87b0)
This spacing is with margin-bottom 10px.
If I remove this margin-bottom in connection-status then it will look like this 
![Screenshot from 2025-01-07 17-02-41](https://github.com/user-attachments/assets/6b4a6494-cacf-4c19-ac60-335e7594c01b)
You can tell me which one looks suitable to you.Thank You and Happy New Year

LianaHarris360 commented:
No need to apologize at all, your efforts are much appreciated! I think the spacing with the margin-bottom of 10px looks the most even. And Happy New Year to you as well!

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

yashhash2 commented:
Hi Liana Is my PR getting merged? and I also wanted to ask about some error whenever I tried to commit using pre-commit hooks of learningequality it showed missing Python.h file are you aware of this error?

LianaHarris360 commented:
Yes, your PR is ready to be merged, I’ll go ahead and approve it. However, it also needs approval from @MisRob, who will be back next week. The actual merging will have to wait until then. Also the failing `Kolibri Build Assets for Pull Request` check is not a result of your changes and is not a blocker for merging the PR.

The missing Python.h file error might be specific to your operating system and the installed python version. There seems to be related issues found [here](https://stackoverflow.com/search?q=missing+Python.h+file&s=1be1b72b-68eb-4d05-82a9-4d8b648de187).

MisRob commented:
Ah sorry @LianaHarris360 I should have dismissed my review. Thank you for taking care of guiding @yashhash2 and @yashhash2 thanks for completing this, seems to be a bit tricky!

Feel free to merge @LianaHarris360 as soon as you feel it's ready.
